[Future Work] Adding LA design for Filesystem Decryption Key Provisioning (#11)
confidential-containers/enclave-cc: docs/design.md

Then like the usual LibOS pattern it will call the container's entry point as though the container
was a process unto itself (rather than where it now sits like a library in the LibOS PAL process).

> ## Local Attestation for FUSE Key Provisioning

Reviewer: Agreed.

> ## Local Attestation for FUSE Key Provisioning

Author: Fixed, thanks.

> ## Local Attestation (v2) for FUSE Key Provisioning

Reviewer: I think v2 is not necessary here

Author: Fixed, thanks.

> ![EnclaveCC_LA.png](images/EnclaveCC_LA.png)

Author: We predetermined the algorithms for use with DH key exchange scheme for simplicity and performance at this stage. I think a TLS like full-fledged negotiation mechanism might be considered to adding if desired.

> ![EnclaveCC_LA.png](images/EnclaveCC_LA.png)

Author: Yes, it could be derived in G6, however, we think it could be better to postpone this operation until it is needed with regard to security, performance and resource consumption, thanks.

> ![EnclaveCC_LA.png](images/EnclaveCC_LA.png)

Reviewer 2: The predetermination is fine. But in this current design, it seems that there are already enough interactions to carry out an algo negotiation (i.e., the performance benefit is not that clear)?

> ![EnclaveCC_LA.png](images/EnclaveCC_LA.png)

Reviewer 2: Thanks for the explanation!

> ![EnclaveCC_LA.png](images/EnclaveCC_LA.png)

Reviewer 2: Got it and good to know. Thanks!

> In most scenarios, it is one-shot transactions used for single FUSE key provisioning. However, the LA protocol can easily be extended to provision multiple keys for more general purposes.

Reviewer: "FUSE Key" is ephemeral and generated by agent-enclave. It is only used one time for app container.

> To achieve this, it is necessary to add a key id to the second request message sent along G_a and app's report, with the ID being used to indicate which key is being requested.

Author: `G_a` is the DH pubkey generated inside app. enclave, thanks.

> To achieve this, it is necessary to add a key id to the second request message sent along G_a and app's report, with the ID being used to indicate which key is being requested.

Author: Yes, it doesn't. I remember a similar issue being raised and discussed at a past weekly enclave-cc meeting, however, we conclude that it won't be added at initial stage as agent enclave has prior knowledge of which key need to be provisioned to app enclave at that phase. @jiazhang0 please chime in if any, Thanks.

> To achieve this, it is necessary to add a key id to the second request message sent along G_a and app's report, with the ID being used to indicate which key is being requested.

Reviewer 2: Thanks! Yep, I understand that the agent enclave needs have prior knowledge - but is this relevant to the proposed protocol (i.e., it's about adding the key ID kind of design instead of the provision steps)? It's just a bit weird that we highlight this in wording but it's not even mentioned in the flow chart. It's also obscure from implementation perspective.

> To achieve this, it is necessary to add a key id to the second request message sent along G_a and app's report, with the ID being used to indicate which key is being requested.

Reviewer: > Thanks! Yep, I understand that the agent enclave needs have prior knowledge - but is this relevant to the proposed protocol (i.e., it's about adding the key ID kind of design instead of the provision steps)? It's just a bit weird that we highlight this in wording but it's not even mentioned in the flow chart. It's also obscure from implementation perspective.

Because whether the key id can be automatically enumerated or not is decided by the caller ttrpc service/client and their contexts. At this moment, the caller is not discussed yet. We hope to merge the protocol definition completed prior to the implementation. By then, we could make the final decision to add the key id into protocol or not. @bigdata-memory I suggest to add such a comment to explain why the key id is not mentioned in flow chart. Thanks!

> To achieve this, it is necessary to add a key id to the second request message sent along G_a and app's report, with the ID being used to indicate which key is being requested.

Reviewer 2: I see, thanks for the context info. Make sense to me.

> In addition to adding this ID, the generated nonce and derived AEK also need to be in the cache on both sides to keep the secure session alive which has been established in initial transaction.

Reviewer 3: What `AEK` stands for?  Can you clarify "secure session" in the context of `ttrpc`.

Author: The `AEK` stands for algorithm encryption key, it is a derived key used for AEAD.
Regarding the secure session, it is tied to `AEK` and has shorter or equal lifetime than `ttrpc` connection. Thanks.

Reviewer 4: Added to the acronym list: https://github.com/confidential-containers/documentation/wiki/Acronyms#aek

Then like the usual LibOS pattern it will call the container's entry point as though the container
was a process unto itself (rather than where it now sits like a library in the LibOS PAL process).

> ## Local Attestation for Filesystem Decryption Key Provisioning

Reviewer 5: I think the best path for enclave-cc right now is to use sealing. There is some discussion in #3 subsequent to the creation of this PR to support the sealing direction.

Please update the heading to reflect that Local Attestation is exploratory or future work.

You will also want to update the design to discuss the policy. The text in the diagram of "check [agent|app] identity against policy" isn't sufficient to evaluate the design here. 

It is also important to define the APIs. 

I would also appreciate some discussion of the use of network interfaces, i.e. are these remotely accessible? How does the app-enclave address the agent-enclave?

Reviewer: I think it would be fine to mark it as exploratory or future work. This doc just gives a design using TEE-agnostic approach to implement an attest channel to transfer a secret. Your questions are reasonable and could be addressed by the real implementation of this LA protocol. The discussion context is here: https://github.com/confidential-containers/enclave-cc/issues/3#issuecomment-1214284752

Reviewer 5: In general, the additional detail I'm requesting is to distinguish the enclave-cc design from local attestation generally.

app's report, with the ID being used to indicate which key is being requested. But whether the key
id can be automatically enumerated or not is decided by the caller `ttrpc` service/client and their
contexts. At this moment, the caller is not discussed yet. We hope to merge the protocol definition
completed prior to the implementation. By then, we could make the final decision to add the key id
into protocol or not.

In addition to adding this ID, the generated nonce and derived `AEAD` keyset expected to be cached
on both sides to keep the secure session alive which has been established in initial transaction.
The session's security context will hold those cached values to ensure AIC triad and freshness for
subsequent key provisioning requests.

Furthermore, agent or app. should set the expiration time of its own security context properly to
minimize risk exposure. Upon closing of underlying `ttrpc` connection, the security context will be
cleared accordingly.

Please note that there is no secure transport configured between `ttrpc` endpoints in our case,
therefore it is necessary to check the integrity and ensure the confidentiality if any of whatever

> they received (e.g., `AEAD`, https://datatracker.ietf.org/doc/html/rfc5116). The `ttrpc` service

Reviewer 5: It is not clear if you are saying that the ttrpc limitation means that there are additional integrity precautions for the enclaves to satisfy (e.g. for the messages in this protocol), or if you are merely saying that we should use AEAD as specified to protect the FDK because AEAD provides integrity.

Author: It is a design choice not the limitation of ttrpc, thanks.

needs to be strictly configured to listen on loopback interface only for security. In addition, the

> identity of app. could also be verified to ensure the provenance of the app. enclave.

Reviewer 5: In this last sentence, I think 'app' and 'agent' got swapped on your latest revision.

But probably this all should be introduced early on and explicitly...
In your initial paragraphs you probably want to say that the agent holds a secret (FDK) and the app wants to authenticate to the agent in order for the agent to release the FDK. The app will prove its identity with a report and the agent will check that against a policy which requires a matching MRSIGNER or MRENCLAVE, possibly including an additional value from KSS. This policy is also really the heart of what this design should explain for enclave-cc.
(i.e. Local attestation as a generic protocol is shown in the linux-sgx examples and the SGX documentation, so discussing things specific to enclave-cc is most helpful for enclave-cc documentation and saves you trouble of rewriting things that may be documented elsewhere. 
Similarly the AEAD algorithm selection should be specified, but if the thinking is just to build on the linux-sgx example it might be better to just reference that.).

In this last sentence I think you meant to offer that the App could also authenticate the Agent. If you want to suggest that, please offer some discussion why this would be beneficial. The only things that come to mind for me are availability attacks which are outside the CC model.
Please also update the diagram to mark the app checking the agent against policy as optional or consider dropping that part.

Author: Thank you for the comments, I have updated the doc and diagram accordingly.

between an Agent-enclave instance and a set of App-enclave instances resided in a same Pod. this
design is a guide for enclave-CC developers to implement this feature.

In this design, the authentication and authorization could be safely simplified in the context of
enclave-CC by verifying and validating the SGX local report with customizable policy rules, e.g.,
matching mrenclave/mrsigner combined with configid.

In most scenarios, it is one-shot transactions used for single filesystem decryption key
provisioning. However, this LA protocol can easily be extended to provision multiple keys for more
general purposes.

To achieve this, it is necessary to add a key id to the second request message sent along G_a and
app's report, with the ID being used to indicate which key is being requested. But whether the key
id can be automatically enumerated or not is decided by the caller ttrpc service/client and their
contexts. At this moment, the caller is not discussed yet. We hope to merge the protocol definition
completed prior to the implementation. By then, we could make the final decision to add the key id
into protocol or not.

> In addition to adding this ID, the generated nonce and derived AEK also expect to be in the cache on

Author: Made it more specific, thanks.

both sides to keep the secure session alive which has been established in initial transaction. The
session's security context will hold those cached values to ensure AIC triad and freshness for
subsequent key provisioning requests.

Furthermore, agent or app. should set the expiration time of its own security context properly to
minimize risk exposure. Upon closing of underlying ttrpc connection, the security context will be
cleared accordingly.

Please note that there is no secure transport configured between ttrpc endpoints in our case,

> therefore the LibOSes need to validate and check integrity of whatever they received. The ttrpc

Author: added a link for reference, thanks.

> therefore the LibOSes need to validate and check integrity of whatever they received. The ttrpc

Reviewer 5: It is not clear if you are saying that the ttrpc limitation means that there are additional integrity precautions for the enclaves to satisfy (e.g. for the messages in this protocol), or if you are merely saying that we should use AEAD as specified to protect the FDK because AEAD provides integrity.

service needs to be strictly configured to listen on loopback interface only for security. With

> regard to the secret in transit, it was designed to be wrapped by a shared key. In addition, the

Reviewer 5: This sentence is unclear. "... it was designed to be wrapped by a shared key." 
Isn't that the whole point of this design? I'm not clear if something additional is intended by this sentence.

Author: removed ambiguous part, thanks.

> identity of agent or app. could also be verified to further ensure the provenance of the app.

Reviewer 5: You are already specifying that the app-enclave identity is explicitly checked by the agent-enclave.
In this sentence do you want to say that the compliment is also possible (i.e. app checks agent)?
If you want to say that please comment to what benefit that provides or risk that would address. It becomes very difficult to have two enclaves check separate MRENCLAVE values.

Author: removed the ambiguous part, thanks.

> enclave container.

Reviewer 5: In the diagram please consistently name the FUSE key. Sometimes it is just mentioned as `key` which can be ambiguous in a key exchange protocol.

can you improve the wording in this step...
<img width="295" alt="image" src="https://user-images.githubusercontent.com/4999494/187075179-bf7bee86-90a9-4551-9919-5b38373d4884.png">

Something like `Encrypt` the fuse key using the AEK rather than `wrap`. 
Add Nonce 1 in `Associated Data`

Author: rephrased it and other associated parts, thanks.

Then like the usual LibOS pattern it will call the container's entry point as though the container
was a process unto itself (rather than where it now sits like a library in the LibOS PAL process).

## [Future Work] Local Attestation for Filesystem Decryption Key Provisioning

![EnclaveCC_LA.png](images/EnclaveCC_LA.png)

A local attestation (LA) feature is needed for authentication/authorization and secret provisioning
between an Agent-enclave instance and a set of App-enclave instances resided in a same Pod. The
agent holds a secret for filesystem decryption and the app wants to authenticate to the agent in

> order for the agent to release the filesystem decryption key (FDK). The app will prove its identity

Reviewer 3: ```suggestion
order for the agent to release the filesystem decryption key (FDK). The app-enclave will prove its identity
```

Author: Thank you so much for those commit suggestions.
It is very helpful to improve the readability. Thanks.

with a report and the agent will check that against a policy which requires a matching `MRSIGNER` or
`MRENCLAVE`, possibly including an additional value from KSS. This design is a guide for enclave-CC
developers to implement this feature.

In this design, we leverages the `AEAD` algorithm for FDK protection over `ttrpc` channel. It is a
recommended cryptographic algorithm in IETF, see [rfc5116](https://datatracker.ietf.org/doc/html/rfc5116).
Regarding the ephemeral Diffie-Hellman key exchange, we chooses the Curve25519, which is recommended
by IETF [rfc8031](https://www.rfc-editor.org/rfc/rfc8031). The design doesn't include the agent
authentication in enclave-CC security context, however, the app can easily add app side user defined
policy to enhance the security configuration.

> In most scenarios, it is one-shot transactions used for single filesystem decryption key (FDK)

Reviewer 3: ```suggestion
In most scenarios, it is one-shot transactions used for single FDK

provisioning. However, this LA protocol can easily be extended to provision multiple keys for more

> general purposes.

Reviewer 3: ```suggestion
general purposes or, for example, multiple application containers in the same pod.
```

To achieve this, it is necessary to add a key id to the second request message sent along G_a and

> app's report, with the ID being used to indicate which key is being requested. But whether the key

Reviewer 3: ```suggestion
app's report, with the key id being used to indicate which key is being requested. But whether the key
```

Reviewer 3: it looks `key id` is commonly used in this paragraph

Author: Agree, thanks.

id can be automatically enumerated or not is decided by the caller `ttrpc` service/client and their
contexts. At this moment, the caller is not discussed yet. We hope to merge the protocol definition
completed prior to the implementation. By then, we could make the final decision to add the key id
into protocol or not.

In addition to adding this ID, the generated nonce and derived `AEAD` keyset expected to be cached
on both sides to keep the secure session alive which has been established in initial transaction.
The session's security context will hold those cached values to ensure AIC triad and freshness for
subsequent key provisioning requests.

Furthermore, agent or app. should set the expiration time of its own security context properly to
minimize risk exposure. Upon closing of underlying `ttrpc` connection, the security context will be
cleared accordingly.

Please note that there is no secure transport configured between `ttrpc` endpoints in our case,

> therefore it is necessary to check the integrity and ensure the confidentiality if any of whatever
> they received using `AEAD`. Furthermore, the `ttrpc` service needs to be strictly configured to

Reviewer 5: I think you just want to call out protecting the FDK specifically. If there are other messages you are trying to cover please add those or describe further what you mean. Thanks for all the updates.
```suggestion
therefore we rely on the confidentiality and integrity measures of `AEAD` to protect the FDK 
over this transport. Furthermore, the `ttrpc` service needs to be strictly configured to
```

Author: Thank you very much for the commit suggestion.
I think there is no other coverage,
so there is no further description for now, thanks.
